Custom Pages for `<UserProfile />` and `<OrganizationProfile />` (#404)
clerkinc/clerk-docs: docs/components/customization/user-profile-custom-pages.mdx

---

> title: Customize User Profile

Reviewer: ```suggestion
title: <UserProfile /> customization
```

> description: How to add custom pages to the `<UserProfile />` component

Reviewer: ```suggestion
description: Learn how to add custom pages and include external links within the navigation sidebar of the <UserProfile /> component.
```

---

> # Custom pages in User Profile
> 
> ## `<UserProfile.Page />`
> 
> The `<UserProfile />` component supports the addition of custom pages. These custom pages can be rendered inside the `<UserProfile />` component and provide a way to incorporate app-specific settings or additional functionality.
> 
> To add a custom page, use the `<UserProfile.Page />` component.

Reviewer: ```suggestion
# `<UserProfile />` customization

The `<UserProfile />` component supports the addition of custom pages and use of external links in the navigation sidebar. 

## `<UserProfile.Page />`

Custom pages can be rendered inside the `<UserProfile />` component and provide a way to incorporate app-specific settings or additional functionality.

To add a custom page to the `<UserProfile />` component, use the `<UserProfile.Page />` component.
```

### Usage

<Tabs items={['Next.js', 'React', 'Remix']}>
    <Tab>

>         `<UserProfile.Page />` can be rendered **only on the client**, and in Next.js app-router projects, they need to be under the `"use client";` flag.

Reviewer: ```suggestion
        `<UserProfile.Page />` can be rendered **only on the client**. For Next.js applications using App Router, the `"use client";` flag must be added.
```

                <UserProfile path="/user-profile" routing="path">
                    <UserProfile.Page label="Custom Page" labelIcon={<CustomIcon />} url="custom-page">
                        <CustomProfilePage />
                    </UserProfile.Page>
                    <UserProfile.Page label="Terms" labelIcon={<CustomIcon />} url="terms">
                        <CustomTerms />
                    </UserProfile.Page>
                </UserProfile>
            );
        }
        ```
    </Tab>
</Tabs>

### Props

All props below are required.

| Name | Type | Description |
| --- | --- | --- |
| `label` | `string` | The name that will be displayed in the navigation sidebar for the custom page. |
| `labelIcon` | `React.ReactElement` | An icon displayed next to the label in the navigation sidebar. |
| `url` | `string` | The path segment that will be used to navigate to the custom page. (e.g. if the `UserProfile` component is rendered at `/user`, then the custom page will be accessed at `/user/{url}` when using path routing) |
| `children` | `React.ReactElement` | The components to be rendered as content inside the custom page. |

## `<UserProfile.Link />`

> In addition to custom pages, you can add external links to the `<UserProfile />` navigation sidebar using the `<UserProfile.Link />` component.

Reviewer: ```suggestion
You can add external links to the `<UserProfile />` navigation sidebar using the `<UserProfile.Link />` component.
```

### Usage

<Tabs items={['Next.js', 'React', 'Remix']}>
    <Tab>

>         `<UserProfile.Link />` can be rendered **only on the client**, and in Next.js app-router projects, they need to be under the `"use client";` flag.

Reviewer: ```suggestion
        `<UserProfile.Link />` can be rendered **only on the client**. For Next.js applications using App Router, the `"use client";` flag must be added.
```

        import { CustomIcon } from "../icons";

        export default function UserProfilePage() {
            return (
                <UserProfile path="/user-profile" routing="path">
                        <UserProfile.Link label="Homepage" labelIcon={<CustomIcon />} url="/home" />
                </UserProfile>
            );
        }
        ```
    </Tab>
</Tabs>

### Props

All props below are required.

| Name | Type | Description |
| --- | --- | --- |
| `label` | `string` | The name that will be displayed in the navigation sidebar for the link.  |
| `labelIcon` | `React.ReactElement` | An icon displayed next to the label in the navigation sidebar. |
| `url` | `string` | The absolute or relative url to navigate to |

## Advanced use cases

### Reordering Default Routes

If you want to reorder the default routes (`Account` and `Security`) in the `UserProfile` navigation sidebar, you can use the `<UserProfile.Page />` component with the `label` prop set to `'account'` or `'security'`. This will target the existing default page and allow you to rearrange it.

### Usage

<Tabs items={['Next.js', 'React', 'Remix']}>
    <Tab>

>         `<UserProfile.Page />` and `<UserProfile.Link />` can be rendered **only on the client**, and in Next.js app-router projects, they need to be under the `"use client";` flag.

Reviewer: ```suggestion
        `<UserProfile.Page />` and `<UserProfile.Link />` can be rendered **only on the client**. For Next.js applications using App Router, the `"use client";` flag must be added.
```

                <UserProfile.Page label="account" />
                <UserProfile.Page label="security" />
            </UserProfile>
        );

        export default UserProfilePage;
        ```
    </Tab>

    <Tab>
        ```tsx filename="routes/user/$.tsx"
        import { UserProfile } from "@clerk/remix";
        import { MyCustomPageContent } from "../components";
        import { CustomIcon, Icon } from "../icons";

        export default function UserProfilePage() {
            return (
                <UserProfile path="/user-profile" routing="path">
                    <UserProfile.Page label="Custom Page" url="custom" labelIcon={<CustomIcon />}>
                        <MyCustomPageContent />
                    </UserProfile.Page>
                    <UserProfile.Link label="Homepage" url="/home" labelIcon={<Icon />} />
                    <UserProfile.Page label="account" />
                    <UserProfile.Page label="security" />
                </UserProfile>
            );
        }
        ```
    </Tab>
</Tabs>

The above example will result in the following order:

1. Custom Page
2. Homepage
3. Account
4. Security

> **Notes**
> 
> - The first page in the list will be rendered under the root path `/` (its `url` will be ignored) and the Clerk pages will be rendered under the path `/account`.
> - The first item in the list cannot be a `<UserProfile.Link />` component.

Reviewer: ```suggestion
It's important to note that the first page in the list will be rendered under the root path `/` (its `url` will be ignored) and the Clerk pages will be rendered under the path `/account`. Also, the first item in the list cannot be a `<UserProfile.Link />` component.
```

### Using custom pages with the `<UserButton />` component

If you are using the `<UserButton />` component with the default props (where the `UserProfile` opens as a modal), then you should also be providing these custom pages as children to the component (using the `<UserButton.UserProfilePage />` and `<UserButton.UserProfileLink />` components respectively).

### Usage

<Tabs items={['Next.js', 'React', 'Remix']}>
    <Tab>

>         `<UserButton.UserProfilePage />` and `<UserButton.UserProfileLink />` can be rendered **only on the client**, and in Next.js app-router projects, they need to be under the `"use client";` flag.

Reviewer: ```suggestion
        `<UserButton.UserProfilePage />` and `<UserButton.UserProfileLink />` can be rendered **only on the client**. For Next.js applications using App Router, the `"use client";` flag must be added.
```
